lib/IDE/Refactoring.cpp

```diff
@@ -6640,17 +6767,80 @@ class AsyncConverter : private SourceEntityWalker {
       llvm_unreachable("Invalid top handler");
     }
 
+    // If ArgNames is not None, then the arguments to the completion handler
+    // were unwrapped using guard-let into variables with these names.
+    Optional<SmallVector<StringRef, 4>> ArgNames;
+
+    if (TopHandler.isAmbiguousCallToParamHandler(CE)) {
+      ArgNames.emplace();
+      // If the completion handler call was ambiguous, then all of its
+      // parameters were most likely of Optional type - otherwise, one of the
+      // parameters would always have to be `nil` or both an error and a success
+      // are passed back.
+      // Since the continuation doesn't take Optional values, we need to unwrap
+      // them.
+      unsigned I = 1;
+      for (auto Arg : Args) {
+        StringRef ArgName;
+        if (auto DRE = dyn_cast<DeclRefExpr>(Arg)) {
+          ArgName = DRE->getDecl()->getBaseIdentifier().str();
+        } else if (auto IIOE = dyn_cast<InjectIntoOptionalExpr>(Arg)) {
+          if (auto DRE = dyn_cast<DeclRefExpr>(IIOE->getSubExpr())) {
+            ArgName = DRE->getDecl()->getBaseIdentifier().str();
+          }
+        }
+        if (ArgName.empty()) {
+          if (Args.size() == 1) {
+            // We only have a single result. 'result' seems a resonable name.
+            ArgName = createUniqueName("result").str();
+          } else {
+            // We are returning a tuple. Name the result elements 'result' +
+            // index in tuple.
+            ArgName = createUniqueName("result" + std::to_string(I)).str();
+          }
+        }
+        ArgNames->push_back(ArgName);
+        OS << tok::kw_guard << ' ' << tok::kw_let << ' ' << ArgName << ' '
+           << tok::equal << ' ';
+        convertNode(Arg, /*StartOverride=*/CE->getArgumentLabelLoc(I),
+                    /*ConvertCalls=*/false);
+        if (auto CE = dyn_cast<CallExpr>(Arg)) {
+          if (CE->hasTrailingClosure()) {
+            // If the argument is a call with trailing closure, the generated
+            // guard statement does not compile.
+            // e.g. 'guard let result1 = value.map { $0 + 1 } else { ... }'
+            // doesn't compile.
+            // Adding a '.self' at the end makes the code compile (although it
+            // will still issue a warning about a trailing closure use inside a
+            // guard condition).
+            OS << tok::period << tok::kw_self;
+          }
+        }
```

Comment:
Is it worth also handling this case for the error argument that's used in the `if let`? Maybe it should be part of `convertNode`?

Reply:
For some reasons, trailing closures are fine in `if let` but not in `guard let`. And since the `.self` isn’t particularly nice if it can be avoided, I’d prefer to leave it as-is.

Reply:
Ah interesting! If anything I would have expected it be the other way around because of the `else` keyword. I agree it's best to leave this as-is then.

Reply:
So would I but 🤷